### DISCUSSION TITLE
I can't stop the job

### DISCUSSION THREAD

**Discussion by nsnhan127** [kelektiv/node-cron#864] [Q&A] [answered]
I created a function to handle click button like this
```javascript
const [published, setPublished] = useState(false);
const handlePublishFlows = async () => {
    const newPublished = !published; 
    setPublished(newPublished); 

    await axios
      .get(`/api/schedule?published=${newPublished}`, {})
      .then((res) => {
        toast.success(res.data.message);
      });
  };
``` 
And this is an api
```javascript
export async function GET(request) {
  const { searchParams } = new URL(request.url);
  const published = searchParams.get("published") === "true";
  console.log(published);
  console.log("Before job instantiation");
  const job = new CronJob("* * * * * *", function () {
    const d = new Date();
    console.log("Every second:", d);
  });
  
  if (published){
    job.start();
    return NextResponse.json({
        message: "Schedule started",
      });
  }
  else{
    job.stop();
    return NextResponse.json({
        message: "Schedule stopped",
      });
  }
}
``` 
My problem is the job still running when i click button to change "published" to "false". How can i fix it ?

**Comment by sheerlox** [comment, marked as answer]
Looks like you're creating a new job on every button click. You would need a way to manage the jobs in your backend so you can find the right one and call the `stop`/`start` methods on it. If you have only one job to manage, maybe moving the job instantiation outside of the `GET` function declaration would do the trick, though not in a very clean way.

**Reply by nsnhan127** [reply]
I tested it outside, it seemed to work. But can I assign the property and call the fireOnTick method like this?
```javascript
import { NextResponse } from "next/server";
import { CronJob } from "cron";
const job = new CronJob();
export async function GET(request) {
  const { params } = new URL(request.url);
  job.cronTime = "* * * * * *";
  job.fireOnTick(()=>{
//do something
})

  if (published) {
    job.start();
    return NextResponse.json({
      message: "Schedule started",
    });
  } else {
    job.stop();
    return NextResponse.json({
      message: "Schedule stopped",
    });
  }
}
``` 
I I tried, but it doesn't seem to work. How can I write an event handler inside the onTick with values taken from the query or body in api?

**Reply by sheerlox** [reply]
You can actually add multiple callbacks (`onTick`s) to the same CronJob. So you could instantiate the `CronJob` with a dummy callback (`() => {}`, since you're required to provide at least one), then from within the `GET` function you could use `job.addCallback(yourFunc)`.

The issue with this approach is that you cannot reset the callbacks, as they are stored in a private property and there is no method on the `CronJob` class to do this (which would make a great feature request, IIRC that was already mentioned some time ago).

One workaround in this case would be to declare your job with `let`, and override it from the `GET` method everytime it's called.

Remember to call `job.stop()` before, because reassigning the variable only means you lose reference to the job, not that it's actually destroyed.

**Reply by nsnhan127** [reply]
So I can't update cron time again, but only update another function in onTick via callback, right?

**Reply by sheerlox** [reply]
You can definitely update the cron time using `setTime`. What you cannot do is override existing callbacks (`onTick`s if you prefer). The only way to do this is to stop the current job, and replace it by a new one. Please let me know if what I'm saying is unclear and I'll try my best to reformulate!

**Reply by nsnhan127** [reply]
thank you very much for supporting
